util/pyclient/bft_metrics_client.py

```diff
@@ -33,22 +35,30 @@ def __exit__(self, *args):
         """context manager method for 'with' statements"""
         self.sock.close()
 
-    def __init__(self, replicas):
-        self.replicas = {}
-        for r in replicas:
-            self.replicas[r.id] = (r.ip, r.port)
+    def __init__(self, replica):
+        self.seq_num = 0
+        self.replica = replica
         self.sock = trio.socket.socket(trio.socket.AF_INET,
                                        trio.socket.SOCK_DGRAM)
+    def _req(self):
+        """Return a get request to the metrics server"""
+        req = struct.pack(HEADER_FMT, REQUEST_TYPE, self.seq_num)
+        self.seq_num += 1
+        return req
 
-    async def get(self, replica):
+    async def get(self):
         """
         Send a get metrics request, retrieve the JSON response, decode it and
         return a map of metrics.
 
         There is no explicit timeout here. Users should call `with
         trio.fail_after as necessary`.
         """
-        await self.sock.sendto(req(), self.replicas[replica])
-        reply, _ = await self.sock.recvfrom(MAX_MSG_SIZE)
-        assert 1 == reply[0]
-        return json.loads(reply[1:])
+        destination = (self.replica.ip,self.replica.port)
+        await self.sock.sendto(self._req(), destination)
+        while True:
+            reply, _ = await self.sock.recvfrom(MAX_MSG_SIZE)
```

Comment:
Do we want to have timeout here?

Reply:
Good catch. But, no, I purposefully didn't put one here. I originally did have a timeout in a prior commit, but I removed it when writing some test code, because it hindered how I wanted to do retries. I wanted retries in the test code (via trio.move_on_after) but not in the metrics client, so I put the timeouts in the test code, and this call in the block that does the trio timeouts. It's a new way of thinking about timeouts, even for me. They aren't attached per operation, but to the surrounding block. It does seem to make some code cleaner, and you don't have to calculate time remaining for overall timeouts which is nice. We may even end up removing them from the udp client to let the caller configure the timeouts in the future, but for now I saw no need to make things more confusing there, especially since retries are more complex.

Reply:
got it.